### PR TITLE
Add new policy [R8] - provide version check macros

### PR DESCRIPTION
**Pull request by balay**
Ref: https://github.com/xsdk-project/xsdk-community-policies/issues/9#issuecomment-501778358

Check https://github.com/xsdk-project/xsdk-community-policies/blob/balay/R8-provide-version-macros/package_policies/R10.md for (display) rendered version of this page

### CODE REVIEW

File: package_policies/R8.md [hunk 1]
```diff
@@ -0,0 +1,7 @@
+**R6.** Each package should provide pre-processor macros for version
```
Comment (mkstoyanov) [inline package_policies/R8.md]: Should this be **R8**?

File: package_policies/R8.md [hunk 2]
```diff
@@ -0,0 +1,7 @@
...
+comparison (for languages that support it - like c/c++/fortran).
+Using these macros, dependent software can potentially be compatible
+with multiple versions of the package.
+
+For example - PACKAGE_VERSION_EQ(), PACKAGE_VERSION_LT(),
```
Comment (mkstoyanov) [inline package_policies/R8.md]: I think it should be enough to provide:
```
#define PACKAGE_VERSION_MAJOR 4
#define PACKAGE_VERSION_MINOR 1
```
Do we need the complex macros `PACKAGE_VERSION_***()`

File: package_policies/R8.md (hunk 1 above)
Comment (balay) [reply package_policies/R8.md]: fixed. Thanks!

File: package_policies/R8.md (hunk 2 above)
Comment (balay) [reply package_policies/R8.md]: I think PACKAGE_VERSION_***() is a better API for users. Otherwise user code would have to construct equivalent complex macros [perhaps multiple times] in the user code. So its best to move this complexity into the package [and not burden the user].

It does requires PACKAGE_VERSION_MAJOR PACKAGE_VERSION_MINOR PACKAGE_VERSION_SUBMINOR and perhaps PACKAGE_VERSION_RELEASE [i.e release or development]

**Comment by balay** [comment]
Example from petsc:
https://bitbucket.org/petsc/petsc/src/master/include/petscversion.h

Example from hdf5 [which provides a subset]
https://bitbucket.hdfgroup.org/projects/HDFFV/repos/hdf5/browse/src/H5public.h

File: package_policies/R8.md (hunk 2 above)
Comment (mkstoyanov) [reply package_policies/R8.md]: The way it is written it is not clear what is actually required to fulfill the policy. The 5 macros are sufficient, but if the package follows good backwards compatibility then a single `PACKAGE_VERSION_GE` is sufficient. If the package changes API on major updates only, then all the user needs is `if (PACKAGE_VERSION_MAJOR > X)` which is just as easy as the `PACKAGE_VERSION_GE`.

In my opinion:

The policy should mandates that packages export variables with compile time information about the version. 
```
#define PACKAGE_VERSION_MAJOR X
#define PACKAGE_VERSION_MINOR Y
#define PACKAGE_VERSION_PATCH Z // if applicable
```
If we mandate additional macros, then `PACKAGE_VERSION_GE` is sufficient.

Comment (balay) [reply package_policies/R8.md]: For one the proposal is currently in the 'recommended' category. If part of it can go into mandatory category - that would be great!

And secondly, I don't think xsdk mandates backward compatibility [and I'm not sure if any package provides it].

And I think its trival to have a complete set of macros - so I don't see it as a burden.. [once a package  commits to having this feature]

Wrt wording - improvements are welcome

**Comment by BarrySmith** [comment]
Providing the macro is a good idea and fairly standard practice. I don't know if I can provide a strong case for needing to provide all of them. @jedbrown

File: package_policies/R8.md (hunk 2 above)
Comment (mkstoyanov) [reply package_policies/R8.md]: Regardless whether the policy is mandatory or recommended, we should all be able to unambiguously answer whether we comply with the policy or not. Also, if a policy becomes mandatory then we should be able to do that with a simple change from **RX** to **MY**.

Here is my take:

For languages that support pre-processor and compile time mactors/variables/methods (e.g., C/C++/Fortran), packages should provide macros that define the version number as well as a method to check whether the current version is at least X.Y.Z.

For example,
```
#define PACKAGE_VERSION_MAJOR X
#define PACKAGE_VERSION_MINOR Y
#define PACKAGE_VERSION_PATCH Z // if applicable

#define PACKAGE_VERSION_GE(Major, Minor, Patch) \
(((PACKAGE_VERSION_MAJOR==Major) && (PACKAGE_VERSION_MINOR==Minor) && (PACKAGE_VERSION_PATCH>=Patch)) || \
        ((PACKAGE_VERSION_MAJOR==Major) && (PACKAGE_VERSION_MINOR>Minor)) || \
        (PACKAGE_VERSION_MAJOR>Major))
```

In C++-17 and above, static-if is also possible
```
namespace Package_Namespace{
constexpr int VERSION_MAJOR = X;
constexpr int VERSION_MINOR = Y;
constexpr int VERSION_PATCH = Z;

auto VERSION_GE(int major, int minor, int patch){
  if (((VERSION_MAJOR==major) && (VERSION_MINOR==minor) && (VERSION_PATCH>=patch)) || 
          ((VERSION_MAJOR==major) && (VERSION_MINOR>minor)) || 
          (VERSION_MAJOR>major)){
    return true;
  }
  return false;
} 
```

*This is sufficient, if packages want to provide more macros there is nothing to prevent that, but there is no need to make it mandatory or even recommended.*

Comment (balos1) [reply package_policies/R8.md]: I agree with @mkstoyanov here. I think that the major, minor, patch macros are sufficient to fulfill the intentions of the policy, which is to provide dependent software the version at compile time. I agree the other macros might be a good practice, but I also agree that a recommended policy should not need to be reworded to become mandatory. Therefore, maybe a better wording would be:

> **R8.** Each package should provide pre-processor macros that allow dependent software to perform compile time comparisons with the current package version (for languages that support it - like c/c++/fortran). Using these macros, dependent software can potentially be compatible with multiple versions of the package.
>
> For example - PACKAGE_VERSION_MAJOR, PACKAGE_VERSION_MINOR, PACKAGE_VERSION_PATCH, PACKAGE_VERSION_GE(major, minor, patch)

This would also allow for packages which dont necessarily follow the semantic versioning format.

Comment (BarrySmith) [reply package_policies/R8.md]: It may seem unnecessary, but I've seem many packages with something like 
#define XXX_VERSION  "x.y.z" 
which is not usefully for machine processing. So I suggest an additional clarifying sentence such as 

Note that providing a version number in an include file as a string does not satisfy this policy since it is difficult to process automatically at compile time.

Comment (mkstoyanov) [reply package_policies/R8.md]: @BarrySmith makes a good point. We should have the note.

Comment (balos1) [reply package_policies/R8.md]: I agree as well.

Comment (balay) [reply package_policies/R8.md]: Hence me mentioning only PACKAGE_VERSION_EQ(), PACKAGE_VERSION_LT() etc API.
We could always suggest an implementation - but I don't think the policy should mandate an implementation.

If policy mandates PACKAGE_VERSION_MAJOR, PACKAGE_VERSION_MINOR, PACKAGE_VERSION_PATCH etc - its mandating a type of implementation - which I don't think is appropriate.

If a package has only  `#define XXX_VERSION "x.y.z"` - then it won;t be able to provide PACKAGE_VERSION_GE() etc API - so won't be compliant.

Also note: different packages might use different names for PACKAGE_VERSION_PATCH.

For ex: PETSc uses PETSC_VERSION_SUBMINOR, hdf5 uses H5_VERS_RELEASE. [edited]

Also - some packages might have x.y notation. Others x.y.z notation - so I avoided saying: 
`PACKAGE_VERSION_EQ(major,minor,subminor)`

**Comment by jedbrown** [comment]
I think `X_VERSION_GE(maj,min,sub)` is the most important in macro guards. If it's the recommended way for clients to guard code, upstream should provide it rather than having each client define something nearly (but often not exactly) equivalent.

One small piece of complexity is whether a package wishes to help pre-release/development users. For example, `PETSC_VERSION_GE` returns true for any development version, thereby allowing dependent packages to make use of pre-release features in a way that will also be correct after the release is made.

File: package_policies/R8.md (hunk 2 above)
Comment (mkstoyanov) [reply package_policies/R8.md]: @balay I think we don't really have a disagreement. My wording does not require `PACKAGE_VERSION_MAJOR` or `_PATCH` but `macros that define the version number`. I also require `a method to check whether the current version is at least X.Y.Z.`, but I don't say what it should be named. The rest is an example, but is not mandatory API.

Under my wording, Petsc is already compliant. Packages that don't have such macros would find it easier to understand to add `PACKAGE_VERSION_<whatever applicable>` and only one `greater than` macro. Under the old wording I was confused on how many macros I should provide.

We all agree that the string "x.y.z" is not a number and is not compliant.

**Review by bartlettroscoe — COMMENTED** [review]
No mention of semver.org?

File: package_policies/R8.md [hunk 3]
```diff
@@ -0,0 +1,7 @@
+**R8.** Each package should provide pre-processor macros for version
```
Comment (balos1) [inline package_policies/R8.md]:
```suggestion
**R8.** Each package should provide pre-processor macros that allow for version
```

**Review by balos1 — DISMISSED** [review]

File: package_policies/R8.md (hunk 3 above)
Comment (balos1) [inline package_policies/R8.md]: This will need to be R9 or R10 now that we have an R8.

Comment (bartlettroscoe) [reply package_policies/R8.md]: Why just "allow"?  With careful usage of `git describe`, you can produce a useful and meaningful version description for version of your code.  (But that gets more complicated for multi-repo projects but you can still do it.)

Comment (mkstoyanov) [reply package_policies/R8.md]: @bartlettroscoe Are you on the call?

The idea is to have some macros, but we leave the package to decide how to do that.

Also, this is at the level of C/C++/Fortran, we don't assume `git`.

Comment (balay) [reply package_policies/R8.md]: fixed now [and squashed/rebased to fix merge conflicts]

Comment (balay) [reply package_policies/R8.md]: @bartlettroscoe - the primary purpose of this policy is to enable usage of type:

https://github.com/trilinos/Trilinos/blob/master/packages/amesos2/src/Amesos2_Superludist_FunctionMap.hpp#L137

**Comment by mkstoyanov** [comment]
The specifics of the implementation are left to the individual package, here are some potential examples. The examples are to be takes as acceptable suggestions and not an exhaustive list of what is acceptable, other solutions are good so long as they provide a programatic way of handling variations of the API.
```
// Example 1: version macros, can be done easily with CMake
#define <PACKAGE>_VERSION_MAJOR 1
#define <PACKAGE>_VERSION_MINOR 2
#define <PACKAGE>_VERSION_PATCH 3
```

```
// Example 2: comparison macros, package version is 1.2.3
#define <PACKAGE>_VERSION_GE(Major, Minor, Patch) \
(((Major == 1) && (Minor == 2) && (Patch <= 3)) || \
        ((Major == 1) && (Minor < 2)) || (Major < 1))

#define <PACKAGE>_VERSION_GT(Major, Minor, Patch) \
(((Major == 1) && (Minor == 2) && (Patch < 3)) || \
        ((Major == 1) && (Minor < 2)) || (Major < 1))

#define <PACKAGE>_VERSION_EQ(Major, Minor, Patch) \
((Major == 1) && (Minor == 2) && (Patch == 3))

#define <PACKAGE>_VERSION_LE(Major, Minor, Patch) \
!<PACKAGE>_VERSION_GT(Major, Minor, Patch)

#define <PACKAGE>_VERSION_LT(Major, Minor, Patch) \
!<PACKAGE>_VERSION_GE(Major, Minor, Patch)
```

```
// Example 3: using a single integer as version
#define <PACKAGE>_VERSION 1002003

// adding optional comparison macros
#define <PACKAGE>_VERSION_GE(Major, Minor, Patch) \
(Major * 10000 + Minor * 100 + Patch <= 1002003)
```

Packages that use C++-17 or newer can alternatively provide `constexpr` variables and methods that can be used in conjunction with `constexpr-if` in place of the `#define` directives.

Implementation that are **NOT** acceptable include:
* providing a string `#define <package>_VERSION "1.2.3"` or even `constexpr` C++-string which is hard to parse
* expecting the user to "know" the version or expecting the user to manually edit source-code
* the user should have access to the macros within the C/C++/Fortrain environment without external tools such as CMake and git, the package can use CMake/git to generate the macros but the user should not need CMake/git to use the macros
* all macro names should be "namespaced" (e.g., use the package as a prefix) to avoid conflicts

**Comment by balay** [comment]
@mkstoyanov I pushed your suggested changes [with some modifications].

All - please review again! The rendered page is available at https://github.com/xsdk-project/xsdk-community-policies/blob/balay/R8-provide-version-macros/package_policies/R10.md

Note: You can provide edits directly via `Add Suggestion` feature at https://github.com/xsdk-project/xsdk-community-policies/pull/16/files

> Packages that use C++-17 or newer can alternatively provide constexpr variables and methods that can be used in conjunction with constexpr-if in place of the #define directives.

Presumably this would also mean that the 'user-code/dependent-library' should also be using c++-17 - and that's a requirement this policy shouldn't impose? So didn't add this part..

File: package_policies/R10.md [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+Using these macros, dependent software can potentially be compatible
+with multiple versions of the package.
+
+Example implementations for a given `package version 1.2.3`
+
+- Example 1: version macros, can be done easily with CMake
+```C
+#define <PACKAGE>_VERSION_MAJOR 1
+#define <PACKAGE>_VERSION_MINOR 2
+#define <PACKAGE>_VERSION_PATCH 3
```
Comment (balay) [inline package_policies/R10.md]: @mkstoyanov I don't understand the CMake example. Does it automatically generate a public  include file - with the corresponding macros?

File: package_policies/R10.md [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+```
+
+- Example 2: comparison macros
```
Comment (balay) [inline package_policies/R10.md]: @mkstoyanov  - how about:

```suggestion
- Example 2: provide package version and comparison macros in a public include file
```

**Comment by shuds13** [comment]
Perhaps for Python we require that version is available via `package.__version__` which is standard practice in python (https://peps.python.org/pep-0008/#module-level-dunder-names).  E.g.

So version can be checked in Python via:

```python
if tuple(numpy.__version__.split(".")) >= ("1", "15", "0"):
```

**Comment by balay** [comment]
Since there is no compile stage with python  - this issue overlaps with (or covered by) https://github.com/xsdk-project/xsdk-community-policies/blob/master/package_policies/M8.md ?

**Comment by shuds13** [comment]
Maybe say, provide 1, or 3 (first line), and, in addition,  we also  encourage comparison macros.

**Comment by mkstoyanov** [comment]
Why don't we just modify R8 to day `runtime and compile time API`, that will cover all cases and we put in the examples for `compile time API`

**Review by mkstoyanov — APPROVED** [review]

**Comment by mkstoyanov** [comment]
> Presumably this would also mean that the 'user-code/dependent-library' should also be using c++-17 - and that's a requirement this policy shouldn't impose? So didn't add this part..

C++ version requirement is usually transitive, i.e., if someone has a hard requirement for C++-17 that's hard to avoid passing that downstream through the headers.

C++ is trying to move away from macros, so I don't want the macros to be required if the C++ native approach is available.

File: package_policies/R10.md (hunk 4 above)
Comment (mkstoyanov) [reply package_policies/R10.md]: You can ignore the CMake comment, I'm just pointing out that this is an easy thing to do.

File: package_policies/R10.md [hunk 6]
```diff
@@ -0,0 +1,56 @@
+**R10.** Each package should provide pre-processor macros that allow for version
+comparison (for languages that support it - like C/C++/Fortran).
+Using these macros, dependent software can potentially be compatible
+with multiple versions of the package.
+
+Example implementations for a given `package version 1.2.3`
+
+- Example 1: version macros, can be done easily with CMake
```
Comment (balay) [inline package_policies/R10.md]: @mkstoyanov 

Perhaps this is what you want to convey

```suggestion
- Example 1: version macros only , via public include files [or can be done easily with CMake]
```

Also since there is emphasis on making the (Example)  comparison macros optional - and current organization 1,2,3 is not conveying that - I'll reorganize them a bit.

Comment (balay) [reply package_policies/R10.md]: @mkstoyanov @v-dobrev @shuds13 I've reorganized the examples a bit - please recheck.

**Comment by balay** [comment]
> C++ version requirement is usually transitive, i.e., if someone has a hard requirement for C++-17 that's hard to avoid passing that downstream through the headers.

@mkstoyanov, my [limited] observation is some pkgs  have  `C++-17` (or such)  required for part of the sources or API - and not all.

One use case for this policy is conditional user code [on the pkg version]. Another use case can be - enforcing a minimum compatible version via configure or cmake tests (on the user pkg side). I guess such tests can be forced to be C++17 code - but I'm not sure what the practice is.

Perhaps you can push an appropriate recommendation for C++17 - and others [more knowledgeable of c++17 interoperability between pkgs] can comment on it.

File: package_policies/R10.md [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+!<PACKAGE>_VERSION_GE(Major, Minor, Patch)
+```
+
+Implementation that are *NOT* acceptable include:
+
+- providing a string #define <PACKAGE>_VERSION "1.2.3"
+
+- expecting the user to "know" the version or expecting the user to manually edit source-code
+
+- requiring the user to use external tools such as CMake and git to generate and access version macros
```
Comment (mkstoyanov) [inline package_policies/R10.md]: It's shorter but I don't know if it conveys what we want.

For example, in Tasmanian, the user must use CMake (or CMake though spack) to build and install Tasmanian and that's how the macros are generated. However, the macros are in the installed headers so the user does NOT need CMake to use Tasmanian or the macros.

Some people may misinterpret this and think that we forbid packages to use CMake to generate macros. But we want to avoid restrictions only on the USER, the package can generate the macros anyway they want.

Comment (balay) [reply package_policies/R10.md]: I pushed an update to clarify further

**Comment by mkstoyanov** [comment]
> @mkstoyanov, my [limited] observation is some pkgs have `C++-17` (or such) required for part of the sources or API - and not all.

I spoke too soon, even C++-17 is not ready to substitute the `#ifdef` macros. Ignore this for now, if necessary, will revisit in few years.

**Comment by balay** [comment]
A note wrt M8 (and potential merge of R10 to M8)  - the motivation for these 2 polices are different.

i.e "having version or commit info, and build info (for all dependent pkgs)  so that the builds can be reproduced later" vs "having version info so conditional user/library code is made possible when package API changes across versions."

And  with python - a single impl of `pkg.__version__ ` feature can satisfy both policies.

**Review by mkstoyanov — APPROVED** [review]

**Review by mkstoyanov — APPROVED** [review]

**Comment by mkstoyanov** [comment]
The motivation is different but the technical aspects are very similar (or identical as in the case of Python). Similarly wrapping macros in run-time API is trivial for C and C++.

We can have one policy with multiple reasons/justifications, although the merge is something to consider down-the-line.

**Comment by shuds13** [comment]
I think we will want to pull in the other PR first, and then modify this to R8, and pull in after (as currently an R8 exists).

We could also change the target to shuds13/policy_review and then I sort out the numbers and the README there, before pulling that in.

**Review by masterleinad — APPROVED** [review]

**Review by shuds13 — APPROVED** [review]

**Review by ulrikeyang — APPROVED** [review]
This is fine if moved to R8 as discussed.

**Review by mgates3 — APPROVED** [review]

**Review by xiaoyeli — APPROVED** [review]

**Review by gardner48 — APPROVED** [review]

**Comment by balay** [comment]
Since PR Authors can't approve [via the approval interface] , I'm adding in my `approval` in this comment section

**Review by hartwiganzt — APPROVED** [review]

**Review by stomov — APPROVED** [review]

**Review by masterleinad — APPROVED** [review]

**Review by stomov — APPROVED** [review]

**Review by luszczek — APPROVED** [review]

**Review by v-dobrev — APPROVED** [review]
